### PR TITLE
Steady state ROM

### PR DESCRIPTION
**Pull request by Relzohery**
This PR adds three main classes for the steady state ROM, namely;
    1. ROMBasis: This gets a user basis form a binary file and stores it in a dense matrix.
    2. OperatotProjection : This takes any matrix type (sparse, dense or shell) and basis and produce the reduced matrix(UTAU)
    3.  ROMSolver: This takes user input, mesh, material, and operator type and solves the eigenvalue problem of the reduced 
        system.

Each class has its unit test.

In the Solver test, I used  1D slab reactor and  tried three operators (Diffusion, EnergyDependent and EnergyIndependent). The Diffusion results seems reasonable.  For the EnergyDpendent operator, while the eigenvalues of the ROM is close to the true value, they don't match the one from the diffusion.

### CODE REVIEW

**Comment by Relzohery** [comment]
@robertsj  I did not solve all the issues yet. Also, I can break this into three smaller PRs, if you want.

**Comment by robertsj** [comment]
I'll take a peak today or tomorrow.  I've got an NRC conference call this afternoon and won't make a meeting with you.  I hope F that I can be in the office.

________________________________

File: src/solvers/rom/ROMSolver.hh [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+   typedef typename Eigensolver<D>::SP_solver            SP_solver;
+   typedef typename Fixed_T::SP_manager                  SP_mg_solver;
+   typedef EnergyDependentEigenLHS<D>                LHS_Operator_T;
+   typedef MGTransportOperator<D>                    RHS_Operator_T;
+   typedef EnergyIndependentEigenOperator<D>             Operator_T;
+
+
+   ROMSolver(SP_input inp, SP_mesh Mesh, SP_material mat);
+
+   void SetBasis();
```
[the remark this answers is not in the document]
Comment (robertsj) [reply src/solvers/rom/ROMSolver.hh:59]: maybe private

File: src/solvers/rom/ROMSolver.hh [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+#define SOLVERS_ROM_ROMSOLVER_HH_
+
+
+#include "callow/vector/Vector.hh"
+#include "callow/matrix/MatrixBase.hh"
+#include "utilities/InputDB.hh"
+#include "material/Material.hh"
+#include "geometry/Mesh.hh"
+#include "FixedSourceManager.hh"
+#include "solvers/EigenvalueManager.hh"
```
Comment (robertsj) [inline src/solvers/rom/ROMSolver.hh:18]: delete

File: src/solvers/rom/ROMSolver.hh [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  SP_input d_input ;
+  SP_mesh d_mesh;
+  SP_material d_mat;
+  SP_matrix  d_A ;
+  SP_matrix d_B;
+  int d_r;
+  int d_n;
+  std::string d_operator;
+  double d_keff;
+  SP_vector x_rom;
```
Comment (robertsj) [inline src/solvers/rom/ROMSolver.hh:83]: define me

File: src/solvers/rom/ROMSolver.cc [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+
+ ROMSolver::Set_FullOperators();
+}
+
+template <class D>
+void ROMSolver<D>::Set_FullOperators()
+{
+  if (d_operator == "diffusion")
+  {
+    SP_lossoperator A (new DiffusionLossOperator(d_input, d_mat, d_mesh, false, 0.0, false, 1.0));
```
Comment (robertsj) [inline src/solvers/rom/ROMSolver.cc:41]: Should able to do 
```suggestion
    d_A  = new DiffusionLossOperator(d_input, d_mat, d_mesh, false, 0.0, false, 1.0);
```

File: src/solvers/rom/ROMSolver.hh [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+   typedef callow::Vector::SP_vector                 SP_vector;
+   typedef detran_utilities::InputDB::SP_input       SP_input;
+   typedef detran_geometry::Mesh::SP_mesh                SP_mesh;
+   typedef detran_material::Material::SP_material        SP_material;
+   typedef detran::DiffusionLossOperator::SP_lossoperator    SP_lossoperator;
+   typedef detran::DiffusionGainOperator::SP_gainoperator    SP_gainoperator;
+   typedef typename Eigensolver<D>::Fixed_T              Fixed_T;
+   typedef typename Eigensolver<D>::SP_solver            SP_solver;
+   typedef typename Fixed_T::SP_manager                  SP_mg_solver;
+   typedef EnergyDependentEigenLHS<D>                LHS_Operator_T;
```
Comment (robertsj) [inline src/solvers/rom/ROMSolver.hh:52]: Maybe name these different so it's easier to see Ax=lambda B x etc

File: src/solvers/rom/ROMSolver.cc [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+  d_keff =  eigensolver->eigenvalue();
+
+  // reconstruct
+  int d_n = d_U->number_rows();
+
+  d_U->multiply(x_rom, sol);
+
+  // correct the direction of the vector if negative
+  for (int i=0; i<d_n; i++)
+  {
```
Comment (robertsj) [inline src/solvers/rom/ROMSolver.cc:122]: maybe try

```suggestion
  if (sol->sum() < 0)
  {
    sol->scale(-1);
  }
```

File: src/solvers/test/test_ROMSolver.cc [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+ input->put<std::string>("equation", "diffusion");
+
+ int n = mesh->number_cells();
+ int r = 5;
+
+// get the basis
+ SP_matrix U;
+ U = new callow::MatrixDense(2*n, 2*r);
+ std::cout << n << "\n";
+ ROMBasis::GetBasis("/home/rabab/opt/detran/source/src/solvers/test/flux_basis_core0_diff", U);
```
Comment (robertsj) [inline src/solvers/test/test_ROMSolver.cc:57]: :(

**Review by robertsj — CHANGES_REQUESTED** [review]
Comments.

File: src/solvers/rom/OperatorProjection.hh [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+
+  private:
+	/// basis rank
+	int d_r;
+	/// operator size
+	int d_n;
+	/// operator to be projected
+	SP_matrix d_A;
+	/// basis onto which the operator is projected
+	SP_matrixDense d_U;
```
Comment (robertsj) [inline src/solvers/rom/OperatorProjection.hh:51]:
```suggestion
	SP_matrixdense d_U;
```

File: src/solvers/rom/ROMBasis.cc [hunk 9]
```diff
@@ -0,0 +1,54 @@
+//----------------------------------*-C++-*-----------------------------------//
+/**
+ *  @file  ROMBasis.cc
+ *  @brief RoMBasis class definition.
```
Comment (robertsj) [inline src/solvers/rom/ROMBasis.cc:4]:
```suggestion
 *  @brief ROMBasis class definition.
```

File: src/solvers/rom/ROMBasis.cc [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  int i;
+  int j;
+  for (int c=0; c<d_r*d_n; c++)
+  {
+    // get the row index
+    i = c/d_r;
+    // get the column index
+    j = c%d_r;
+
+    U->insert(i, j, B[c]);
```
Comment (robertsj) [inline src/solvers/rom/ROMBasis.cc:49]: is this assembled after?

File: src/solvers/rom/ROMSolver.hh [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+
+#include "callow/vector/Vector.hh"
+#include "callow/matrix/MatrixBase.hh"
+#include "utilities/InputDB.hh"
+#include "material/Material.hh"
+#include "geometry/Mesh.hh"
+#include "FixedSourceManager.hh"
+#include "solvers/EigenvalueManager.hh"
+#include "solvers/mg/DiffusionLossOperator.hh"
```
Comment (robertsj) [inline src/solvers/rom/ROMSolver.hh:20]:
```suggestion
```

**Review by robertsj — APPROVED** [review]

**Pull request merged**